File: implementations/node/src/url.ts
import { UsageError } from "./errors.js";

const DEFAULT_ROUTE_BASE = "/devtools";

export function normalizeDeviceUrl(input: string): string {
  const trimmed = input.trim();
  if (!trimmed) {
    throw new UsageError("Device host cannot be empty.");
  }
  if (trimmed.includes("\0")) {
    throw new UsageError("Device host contains an invalid NUL character.");
  }

  const withScheme = /^[a-z][a-z\d+.-]*:\/\//i.test(trimmed) ? trimmed : `http://${trimmed}`;
  let url: URL;
  try {
    url = new URL(withScheme);
  } catch (error) {
    throw new UsageError(`Invalid device host: ${input}`, { cause: error });
  }

  if (url.protocol !== "http:" && url.protocol !== "https:") {
    throw new UsageError(`Unsupported device URL scheme: ${url.protocol}`);
  }
  if (url.username || url.password) {
    throw new UsageError("Credentials are not allowed in the device URL.");
  }
  if (url.search || url.hash) {
    throw new UsageError("Device URL must not contain a query string or fragment.");
  }

  let pathname = url.pathname.replace(/\/+$/, "");
  if (!pathname) pathname = DEFAULT_ROUTE_BASE;
  if (pathname.endsWith("/api")) pathname = pathname.slice(0, -4);
  if (pathname !== DEFAULT_ROUTE_BASE) {
    throw new UsageError(`Device URL path must be ${DEFAULT_ROUTE_BASE} or ${DEFAULT_ROUTE_BASE}/api.`);
  }

  url.pathname = DEFAULT_ROUTE_BASE;
  return url.toString().replace(/\/$/, "");
}
export function apiUrl(baseUrl: string, route: string, query: Record<string, string | number | boolean | undefined> = {}): URL {
  const cleanRoute = route.replace(/^\/+/, "");
  const url = new URL(`${baseUrl}/api/${cleanRoute}`);
  for (const [key, value] of Object.entries(query)) {
    if (value !== undefined) url.searchParams.set(key, String(value));
  }
  return url;
}

File: implementations/rust/src/url.rs
use url::Url;

use crate::error::{CubicError, Result};

pub fn normalize_device_url(value: &str) -> Result<String> {
    let trimmed = value.trim();
    if trimmed.is_empty() {
        return Err(CubicError::usage("Device host cannot be empty."));
    }
    if trimmed.contains('\0') {
        return Err(CubicError::usage(
            "Device host contains an invalid NUL character.",
        ));
    }
    let candidate = if trimmed.contains("://") {
        trimmed.to_owned()
    } else {
        format!("http://{trimmed}")
    };
    let mut url = Url::parse(&candidate)
        .map_err(|_| CubicError::usage(format!("Invalid device host: {value}")))?;
    if url.scheme() != "http" && url.scheme() != "https" {
        return Err(CubicError::usage(format!(
            "Unsupported device URL scheme: {}",
            url.scheme()
        )));
    }
    if url.host_str().is_none() {
        return Err(CubicError::usage(format!("Invalid device host: {value}")));
    }
    if !url.username().is_empty() || url.password().is_some() {
        return Err(CubicError::usage(
            "Credentials are not allowed in the device URL.",
        ));
    }
    if url.query().is_some() || url.fragment().is_some() {
        return Err(CubicError::usage(
            "Device URL must not contain a query string or fragment.",
        ));
    }
    match url.path().trim_end_matches('/') {
        "" | "/devtools" | "/devtools/api" => url.set_path("/devtools"),
        _ => {
            return Err(CubicError::usage(
                "Device URL path must be /devtools or /devtools/api.",
            ));
        }
    }
    Ok(url.to_string().trim_end_matches('/').to_owned())
}

pub fn api_url(base_url: &str, route: &str, query: &[(&str, String)]) -> Result<String> {
    let clean_route = route.trim_start_matches('/');
    let mut url = Url::parse(&format!("{base_url}/api/{clean_route}"))
        .map_err(|error| CubicError::new(format!("Invalid API URL: {error}"), "INVALID_URL"))?;
    {
        let mut pairs = url.query_pairs_mut();
        for (key, value) in query {
            pairs.append_pair(key, value);
        }
    }
    Ok(url.into())
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn normalizes_and_rejects_urls() {
        assert_eq!(
            normalize_device_url("192.0.2.42").unwrap(),
            "http://192.0.2.42/devtools"
        );
        assert_eq!(
            normalize_device_url("http://host/devtools/api/").unwrap(),
            "http://host/devtools"
        );
        for input in [
            "",
            "ftp://host",
            "http://user:pass@host",
            "http://host/other",
            "http://host?x=1",
        ] {
            assert!(normalize_device_url(input).is_err(), "{input}");
        }
    }
}
